### ISSUE TITLE
Self is not rejected when used as a PEP 695 type parameter bound

### ISSUE THREAD

**Issue by EmmanuelNiyonshuti** [python/mypy#21960]
### Bug Report

Self is only valid in the specific locations listed in [PEP 673](https://peps.python.org/pep-0673/#valid-locations-for-self) and Mypy already rejects several invalid locations.
See: https://github.com/python/mypy/blob/ae39cdb2fba8908eeff6ec8d6b88879c62495fcc/test-data/unit/check-selftype.test#L1480

It seems one location wasn't covered: using Self as the bound (or constraint) of a PEP 695 type parameter.

### To Reproduce:
```python
from typing import Self
class Foo:
      def fails[T: Self](self: T) -> None: pass
```
### Expected Behavior

Mypy should reject `Self` here with an error.

From a quick look at other type checkers: pyright and ty treat it as an invalid/generic TypeVar bound (`TypeVar upper bound cannot be generic`), while zuban rejects it with the same message mypy already uses elsewhere,
```Self type is only allowed in annotations within class definition```.

### Actual Behavior
No error is reported.

### Your Environment

Mypy version: https://github.com/python/mypy/commit/ae39cdb2fba8908eeff6ec8d6b88879c62495fcc
Python version: 3.12

TLDR; originally spotted while working on #21928 See: https://github.com/python/mypy/pull/21928#issuecomment-5557910054
